lib/zhuyin.ex

```diff
@@ -0,0 +1,376 @@
+    "ㄨ" => "wu",
+    "ㄩ" => "yu",
+    "ㄧㄚ" => "ya",
+    "ㄨㄚ" => "wa",
+    "ㄧㄥ" => "ying",
+    "ㄧㄤ" => "yang",
+    "ㄧㄝ" => "ye",
+    "ㄨㄛ" => "wo",
+    "ㄨㄥ" => "weng",
+    "ㄨㄤ" => "wang",
+    "ㄧㄠ" => "yao",
+    "ㄨㄞ" => "wai",
+    "ㄩㄝ" => "yue",
+    "ㄩㄥ" => "yong",
+    "ㄧㄡ" => "you",
+    "ㄨㄟ" => "wei",
+    "ㄧㄢ" => "yan",
+    "ㄨㄢ" => "wan",
+    "ㄩㄢ" => "yuan",
+    "ㄧㄣ" => "yin",
+    "ㄨㄣ" => "wen",
+    "ㄩㄣ" => "yun",
+    # Technically standalone initials. Parsed as standalone finals because it's easier to deal with
+    "ㄓ" => "zhi",
+    "ㄔ" => "chi",
+    "ㄕ" => "shi",
+    "ㄖ" => "ri",
+    "ㄗ" => "zi",
+    "ㄘ" => "ci",
+    "ㄙ" => "si",
+    # Standalone finals that are the same in Pinyin as when combined with an initial
+    "ㄦ" => "er",
+    "ㄢ" => "an"
+  }
+
+  @finals %{
+    "ㄧ" => "i",
+    "ㄨ" => "u",
+    "ㄩ" => "v",
+    "ㄚ" => "a",
+    "ㄛ" => "o",
+    "ㄜ" => "e",
+    "ㄝ" => "e",
+    "ㄞ" => "ai",
+    "ㄟ" => "ei",
+    "ㄠ" => "ao",
+    "ㄡ" => "ou",
+    "ㄢ" => "an",
+    "ㄣ" => "en",
+    "ㄤ" => "ang",
+    "ㄥ" => "eng",
+    "ㄦ" => "er",
+    "ㄧㄚ" => "ia",
+    "ㄨㄚ" => "ua",
+    "ㄧㄥ" => "ing",
+    "ㄧㄤ" => "iang",
+    "ㄧㄝ" => "ie",
+    "ㄨㄛ" => "uo",
+    "ㄨㄥ" => "ong",
+    "ㄨㄤ" => "uang",
+    "ㄧㄠ" => "iao",
+    "ㄨㄞ" => "uai",
+    "ㄩㄝ" => "ve",
+    "ㄩㄥ" => "iong",
+    "ㄧㄡ" => "iu",
+    "ㄨㄟ" => "ui",
+    "ㄧㄢ" => "ian",
+    "ㄨㄢ" => "uan",
+    "ㄩㄢ" => "van",
+    "ㄧㄣ" => "in",
+    "ㄨㄣ" => "un",
+    "ㄩㄣ" => "vn"
+  }
+
+  @doc """
+  Create a pinyin struct from a zhuyin struct or list.
+
+  ## Examples
+
+      iex> Zhuyin.to_pinyin(~z/ㄋㄧˇㄏㄠˇ/)
+      ~p/nǐhǎo/
+
+      iex> Zhuyin.to_pinyin(%Zhuyin{initial: "ㄋ", final: "ㄧ", tone: 3})
+      %Pinyin{initial: "n", final: "i", tone: 3}
+
+  """
+  @spec to_pinyin(t() | zhuyin_list()) :: Pinyin.t() | Pinyin.pinyin_list()
+  # Special case for this final and tone combination
+  def to_pinyin(%Zhuyin{initial: "", final: "ㄦ", tone: 0}) do
+    %Pinyin{initial: "", final: "r", tone: "0"}
+  end
+
+  def to_pinyin(zhuyin = %Zhuyin{}) do
+    pinyin_initial = Map.get(@initials, zhuyin.initial, "")
+
+    pinyin_final =
+      if zhuyin.initial == "" do
+        @standalone_finals[zhuyin.final]
+      else
+        @finals[zhuyin.final]
+      end
+
+    %Pinyin{initial: pinyin_initial, final: pinyin_final, tone: zhuyin.tone}
+  end
+
+  def to_pinyin(list) when is_list(list) do
+    list
+    |> Enum.map(fn
+      z = %Zhuyin{} -> Zhuyin.to_pinyin(z)
+      str when is_binary(str) -> str
+    end)
+  end
+
+  @doc """
+  Create a pinyin struct from a zhuyin struct or list.
+
+  ## Examples
+
+      iex> Zhuyin.from_pinyin(~p/nǐhǎo/)
+      ~z/ㄋㄧˇㄏㄠˇ/
+
+      iex> Zhuyin.from_pinyin(%Pinyin{initial: "n", final: "i", tone: 3})
+      %Zhuyin{initial: "ㄋ", final: "ㄧ", tone: 3}
+
+  """
+  @spec from_pinyin(Pinyin.t() | Pinyin.pinyin_list()) :: t()
+  def from_pinyin(pinyin = %Pinyin{}) do
+    initial_map = Map.new(@initials, fn {key, val} -> {val, key} end)
```

Comment:
This will create a new map every time this function is called. Perhaps you can create this map at compile time (`@foo Map.new(@initials, fn {key, val} -> {val, key} end)`) to avoid this.

Reply:
Good idea! Done.